# Mental Health Dataset

# Data Analysis by Praveen Sagar K V J

In [10]:
#Importing all the required libraries
import numpy as np
import pandas as pd

In [11]:
#reading the dataset from excel file
df=pd.read_csv('C:\\Users\\ARUN PRAVEEN\\Desktop\\MENTAL HEALTH DATASET.csv')

#The dataset has 27 columns but it is much wider then the notebook so we can display all the dataframe columns by the following code so that we won't miss middle columns
pd.set_option('display.max_columns',50)

In [12]:
df.head(5)

Timestamp  Age  Gender         Country state self_employed  \
0  2014-08-27 11:29:31   37  Female   United States    IL           NaN   
1  2014-08-27 11:29:37   44       M   United States    IN           NaN   
2  2014-08-27 11:29:44   32    Male          Canada   NaN           NaN   
3  2014-08-27 11:29:46   31    Male  United Kingdom   NaN           NaN   
4  2014-08-27 11:30:22   31    Male   United States    TX           NaN   

  family_history treatment work_interfere    no_employees remote_work  \
0             No       Yes          Often            6-25          No   
1             No        No         Rarely  More than 1000          No   
2             No        No         Rarely            6-25          No   
3            Yes       Yes          Often          26-100          No   
4             No        No          Never         100-500         Yes   

  tech_company    benefits care_options wellness_program   seek_help  \
0          Yes         Yes     Not sure               No         Yes   
1           No  Don't know           No       Don't know  Don't know   
2          Yes          No           No               No          No   
3          Yes          No          Yes               No          No   
4          Yes         Yes           No       Don't know  Don't know   

    anonymity               leave mental_health_consequence  \
0         Yes       Somewhat easy                        No   
1  Don't know          Don't know                     Maybe   
2  Don't know  Somewhat difficult                        No   
3          No  Somewhat difficult                       Yes   
4  Don't know          Don't know                        No   

  phys_health_consequence     coworkers supervisor mental_health_interview  \
0                      No  Some of them        Yes                      No   
1                      No            No         No                      No   
2                      No           Yes        Yes                     Yes   
3                     Yes  Some of them         No                   Maybe   
4                      No  Some of them        Yes                     Yes   

  phys_health_interview mental_vs_physical obs_consequence comments  
0                 Maybe                Yes              No      NaN  
1                    No         Don't know              No      NaN  
2                   Yes                 No              No      NaN  
3                 Maybe                 No             Yes      NaN  
4                   Yes         Don't know              No      NaN

In [13]:
#Now I'm displaying the Non null count and data types for each of the variable
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1259 entries, 0 to 1258
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   Timestamp                  1259 non-null   object
 1   Age                        1259 non-null   int64 
 2   Gender                     1259 non-null   object
 3   Country                    1259 non-null   object
 4   state                      744 non-null    object
 5   self_employed              1241 non-null   object
 6   family_history             1259 non-null   object
 7   treatment                  1259 non-null   object
 8   work_interfere             995 non-null    object
 9   no_employees               1259 non-null   object
 10  remote_work                1259 non-null   object
 11  tech_company               1259 non-null   object
 12  benefits                   1259 non-null   object
 13  care_options               1259 non-null   object
 14  wellness

from the info we can say that only few columns ("state","work_interfere" ,"comments") has null values 

In [14]:
#Now we can summarize number of missing values for each of the columns and the percentage of it
missing_values_per_column = df.isnull().sum().sort_values(ascending=False)
percentage_missing_values = 100*(df.isnull().sum()/df.isnull().count()).sort_values(ascending=False)
missing = pd.concat([missing_values_per_column, percentage_missing_values], 
                         axis=1, keys=['count of missing values', 'Percentage of missing values'])

ms= 'Percentage of missing values'
missing[ms] = pd.Series(["{0:.2f}%".format(val) for val in missing[ms]],index = missing.index)
missing


count of missing values  \
comments                                      1095   
state                                          515   
work_interfere                                 264   
self_employed                                   18   
benefits                                         0   
Age                                              0   
Gender                                           0   
Country                                          0   
family_history                                   0   
treatment                                        0   
no_employees                                     0   
remote_work                                      0   
tech_company                                     0   
care_options                                     0   
obs_consequence                                  0   
wellness_program                                 0   
seek_help                                        0   
anonymity                                        0   
leave                                            0   
mental_health_consequence                        0   
phys_health_consequence                          0   
coworkers                                        0   
supervisor                                       0   
mental_health_interview                          0   
phys_health_interview                            0   
mental_vs_physical                               0   
Timestamp                                        0   

                          Percentage of missing values  
comments                                        86.97%  
state                                           40.91%  
work_interfere                                  20.97%  
self_employed                                    1.43%  
benefits                                         0.00%  
Age                                              0.00%  
Gender                                           0.00%  
Country                                          0.00%  
family_history                                   0.00%  
treatment                                        0.00%  
no_employees                                     0.00%  
remote_work                                      0.00%  
tech_company                                     0.00%  
care_options                                     0.00%  
obs_consequence                                  0.00%  
wellness_program                                 0.00%  
seek_help                                        0.00%  
anonymity                                        0.00%  
leave                                            0.00%  
mental_health_consequence                        0.00%  
phys_health_consequence                          0.00%  
coworkers                                        0.00%  
supervisor                                       0.00%  
mental_health_interview                          0.00%  
phys_health_interview                            0.00%  
mental_vs_physical                               0.00%  
Timestamp                                        0.00%

In [15]:
# Explore the numeric variable Age to confirm whether all values are within a reasonable range and if any are NaN.
print("Age Analysis")
print("Minimum value: ", df["Age"].min())
print("Maximum value: ", df["Age"].max())
print("How many values are NaN?: ", pd.isnull(df['Age']).sum())


Age Analysis
Minimum value:  -1726
Maximum value:  99999999999
How many values are NaN?:  0


In [16]:
# Now we can analyse variable Gender, which appears not to be standardized

# Count the number of distinct responses and list them:
print("Count of distinct responses for Gender:", len(set(df['Gender'])))
print("Distinct responses for Gender:", set(df['Gender']))

Count of distinct responses for Gender: 49
Distinct responses for Gender: {'Male', 'male', 'Femake', 'femail', 'Male (CIS)', 'Cis Man', 'female', 'queer', 'M', 'Cis Female', 'something kinda male?', 'fluid', 'male leaning androgynous', 'Male ', 'Trans woman', 'non-binary', 'Nah', 'Neuter', 'Female', 'maile', 'Genderqueer', 'Agender', 'Male-ish', 'f', 'woman', 'Female (trans)', 'Androgyne', 'Man', 'Cis Male', 'F', 'p', 'Female (cis)', 'Enby', 'Mal', 'Woman', 'Guy (-ish) ^_^', 'Trans-female', 'Malr', 'Mail', 'All', 'queer/she/they', 'cis-female/femme', 'A little about you', 'ostensibly male, unsure what that really means', 'msle', 'cis male', 'm', 'Make', 'Female '}


My further observation on the dataset

I want to
1) standardize all column header to lower case to avoid typos
2) dtype conversion of Timestamp(object) to datetime
3) remove AGE outliers
4) group GENDER into standardized categories

In [17]:
#standardizing all column header to lower case to avoid typos
df.columns = map(str.lower, df.columns)
df.columns

Index(['timestamp', 'age', 'gender', 'country', 'state', 'self_employed',
       'family_history', 'treatment', 'work_interfere', 'no_employees',
       'remote_work', 'tech_company', 'benefits', 'care_options',
       'wellness_program', 'seek_help', 'anonymity', 'leave',
       'mental_health_consequence', 'phys_health_consequence', 'coworkers',
       'supervisor', 'mental_health_interview', 'phys_health_interview',
       'mental_vs_physical', 'obs_consequence', 'comments'],
      dtype='object')

In [18]:
#dtype conversion of Timestamp(object) to datetime
df['timestamp'] = pd.to_datetime(df['timestamp'])
df.dtypes


timestamp                    datetime64[ns]
age                                   int64
gender                               object
country                              object
state                                object
self_employed                        object
family_history                       object
treatment                            object
work_interfere                       object
no_employees                         object
remote_work                          object
tech_company                         object
benefits                             object
care_options                         object
wellness_program                     object
seek_help                            object
anonymity                            object
leave                                object
mental_health_consequence            object
phys_health_consequence              object
coworkers                            object
supervisor                           object
mental_health_interview         

In [19]:
#removing 'AGE' outliers
# Create a new column "updated_age" that replaces out-of-range ages with Median of the Age column
# The oldest living person on record lived to be 122 years, 164 days so I ll keep age limit as 122
def update_age(age):
    if age>=0 and age<=122:
        return age
    else:
        return df['age'].median()
df['updated_age'] = df['age'].apply(update_age)

# Checking out the new column and making sure it looks right

print("New updated Age")
print("Minimum value: ", df["updated_age"].min())
print("Maximum value: ", df["updated_age"].max())
print("How many values are NaN?: ", pd.isnull(df['updated_age']).sum())
print("Frequency table for updated_age:")
print(df["updated_age"].value_counts().sort_index(0))
df['updated_age'] = df['updated_age'].astype(int)

New updated Age
Minimum value:  5.0
Maximum value:  72.0
How many values are NaN?:  0
Frequency table for updated_age:
5.0      1
8.0      1
11.0     1
18.0     7
19.0     9
20.0     6
21.0    16
22.0    21
23.0    51
24.0    46
25.0    61
26.0    75
27.0    71
28.0    68
29.0    85
30.0    63
31.0    72
32.0    82
33.0    70
34.0    65
35.0    55
36.0    37
37.0    43
38.0    39
39.0    33
40.0    33
41.0    21
42.0    20
43.0    28
44.0    11
45.0    12
46.0    12
47.0     2
48.0     6
49.0     4
50.0     6
51.0     5
53.0     1
54.0     3
55.0     3
56.0     4
57.0     3
58.0     1
60.0     2
61.0     1
62.0     1
65.0     1
72.0     1
Name: updated_age, dtype: int64


In [20]:
#checking the updated_age
df['updated_age'].isna().value_counts()

False    1259
Name: updated_age, dtype: int64

In [21]:
#grouping GENDER into standardized categories

I have decided to keep gender as the following categories for the better analysis
*Male(cis)
*Male(trans)
*Female(cis)
*Female(trans)
*Other

In [22]:

# Standardize the capitalization in the responses by making all characters lower case
df['gender'] = df['gender'].str.lower()

# Make a copy of the column to preserve the original data. I will work with the new column going forward.
df['gender_new'] = df['gender']

# Assign each type of response to one of the five categories

male = ["male", "m", "male-ish", "maile", "mal", "male (cis)", "make", "male ", "man", "msle", "mail", "malr","cis man", "cis male"]
trans_male = [None]
trans_female = ["trans-female",  "trans woman", "female (trans)"]
female = ["cis female", "f", "female", "woman",  "femake", "female ","cis-female/femme", "female (cis)", "femail"]
other = ["non-binary", "nah", "all", "enby", "fluid", "genderqueer", "androgyne", "agender", "male leaning androgynous", "guy (-ish) ^_^", "neuter", "queer", "ostensibly male, unsure what that really means", "queer/she/they", "something kinda male?"]

df['gender_new'] = df['gender_new'].apply(lambda x:"Male (Cis)" if x in male else x)
df['gender_new'] = df['gender_new'].apply(lambda x:"Female (Cis)" if x in female else x)
df['gender_new'] = df['gender_new'].apply(lambda x:"Female (Trans)" if x in trans_female else x)
df['gender_new'] = df['gender_new'].apply(lambda x:"Male (Trans)" if x in trans_male else x)
df['gender_new'] = df['gender_new'].apply(lambda x:"Other" if x in other else x)
df.drop(df[df.gender_new == 'p'].index, inplace=True)
df.drop(df[df.gender_new == 'a little about you'].index, inplace=True)

# Make a crosstab to view the count for each of the new categories
print("Gender:")
print("How many values are NaN?: ", pd.isnull(df['gender_new']).sum())
print("")
print("Frequency table for gender_new:\n", df["gender_new"].value_counts().sort_index(0))
print("")

# Confirm that no entries were missed in the sorting above: display the size of the old and new variables, and of the entire dataset
print("If we didn't miss any entries, these numbers will be the same:")
print("gender =", len(df['gender']), "values")
print("gender_new =", len(df['gender_new']), "values")
print("Dataset Entries =", len(df), "values")

Gender:
How many values are NaN?:  0

Frequency table for gender_new:
 Female (Cis)      247
Female (Trans)      4
Male (Cis)        991
Other              15
Name: gender_new, dtype: int64

If we didn't miss any entries, these numbers will be the same:
gender = 1257 values
gender_new = 1257 values
Dataset Entries = 1257 values


In [23]:
#dropping age columns because we have updated_age

df=df.drop(columns='age')

In [24]:
#dropping gender column
df=df.drop(columns='gender')

In [25]:
df.head()

timestamp         country state self_employed family_history  \
0 2014-08-27 11:29:31   United States    IL           NaN             No   
1 2014-08-27 11:29:37   United States    IN           NaN             No   
2 2014-08-27 11:29:44          Canada   NaN           NaN             No   
3 2014-08-27 11:29:46  United Kingdom   NaN           NaN            Yes   
4 2014-08-27 11:30:22   United States    TX           NaN             No   

  treatment work_interfere    no_employees remote_work tech_company  \
0       Yes          Often            6-25          No          Yes   
1        No         Rarely  More than 1000          No           No   
2        No         Rarely            6-25          No          Yes   
3       Yes          Often          26-100          No          Yes   
4        No          Never         100-500         Yes          Yes   

     benefits care_options wellness_program   seek_help   anonymity  \
0         Yes     Not sure               No         Yes         Yes   
1  Don't know           No       Don't know  Don't know  Don't know   
2          No           No               No          No  Don't know   
3          No          Yes               No          No          No   
4         Yes           No       Don't know  Don't know  Don't know   

                leave mental_health_consequence phys_health_consequence  \
0       Somewhat easy                        No                      No   
1          Don't know                     Maybe                      No   
2  Somewhat difficult                        No                      No   
3  Somewhat difficult                       Yes                     Yes   
4          Don't know                        No                      No   

      coworkers supervisor mental_health_interview phys_health_interview  \
0  Some of them        Yes                      No                 Maybe   
1            No         No                      No                    No   
2           Yes        Yes                     Yes                   Yes   
3  Some of them         No                   Maybe                 Maybe   
4  Some of them        Yes                     Yes                   Yes   

  mental_vs_physical obs_consequence comments  updated_age    gender_new  
0                Yes              No      NaN           37  Female (Cis)  
1         Don't know              No      NaN           44    Male (Cis)  
2                 No              No      NaN           32    Male (Cis)  
3                 No             Yes      NaN           31    Male (Cis)  
4         Don't know              No      NaN           31    Male (Cis)

In [29]:
#Now filling the Null values of work_interfere with "Don't know" (assumption)
df['work_interfere']=df['work_interfere'].fillna("Don't know")

In [30]:
df['work_interfere'].value_counts()

Sometimes     465
Don\t know    264
Never         213
Rarely        173
Often         142
Name: work_interfere, dtype: int64

In [31]:
#There were a few self-employed values listed as NaN, so I assumed that these people were self-employed.
df['self_employed'].fillna('Yes', inplace=True)


In [32]:
df['self_employed'].value_counts()

No     1095
Yes     162
Name: self_employed, dtype: int64

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1257 entries, 0 to 1258
Data columns (total 27 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   timestamp                  1257 non-null   datetime64[ns]
 1   country                    1257 non-null   object        
 2   state                      742 non-null    object        
 3   self_employed              1257 non-null   object        
 4   family_history             1257 non-null   object        
 5   treatment                  1257 non-null   object        
 6   work_interfere             1257 non-null   object        
 7   no_employees               1257 non-null   object        
 8   remote_work                1257 non-null   object        
 9   tech_company               1257 non-null   object        
 10  benefits                   1257 non-null   object        
 11  care_options               1257 non-null   object        
 12  wellne

In [34]:
#we can drop comments column because it has more than 80% missing values
df=df.drop(columns='comments')

In [35]:
#Analysis
df['tech_company'].value_counts()

Yes    1029
No      228
Name: tech_company, dtype: int64

In [36]:
df['no_employees'].value_counts()

6-25              290
26-100            289
More than 1000    282
100-500           176
1-5               160
500-1000           60
Name: no_employees, dtype: int64

In [37]:
df['family_history'].value_counts()

No     767
Yes    490
Name: family_history, dtype: int64

In [38]:
df['treatment'].value_counts()

Yes    635
No     622
Name: treatment, dtype: int64


By observation, there were 1257 people surveyed. Out of this population, the dominant age group was late twenties to early thirties, with the majority being male. About half of those surveyed sought treatment for mental health, while slightly more than half did not have a family history of mental health problems.

# discussing a physical / mental health issue with your employer would have negative consequences?

In [39]:
phys_health_consequences = df.groupby('phys_health_consequence').count()['updated_age']
mental_health_consequences = df.groupby('mental_health_consequence').count()['updated_age']

In [40]:
phys_health_consequences

phys_health_consequence
Maybe    273
No       925
Yes       59
Name: updated_age, dtype: int64

In [41]:
mental_health_consequences 

mental_health_consequence
Maybe    477
No       490
Yes      290
Name: updated_age, dtype: int64

# willing to discuss a mental health issue with your coworkers or direct supervisor(s)?

In [42]:
df['coworkers'].value_counts()

Some of them    774
No              260
Yes             223
Name: coworkers, dtype: int64

In [43]:
df['supervisor'].value_counts()

Yes             514
No              393
Some of them    350
Name: supervisor, dtype: int64

more people would say yes to discussing mental health with their supervisors compared to their coworkers.

In [44]:
df['country'].value_counts()

United States             750
United Kingdom            185
Canada                     72
Germany                    45
Ireland                    27
Netherlands                27
Australia                  21
France                     13
India                      10
New Zealand                 8
Switzerland                 7
Italy                       7
Sweden                      7
Poland                      7
South Africa                6
Belgium                     6
Brazil                      6
Israel                      5
Singapore                   4
Bulgaria                    4
Mexico                      3
Finland                     3
Austria                     3
Russia                      3
Portugal                    2
Greece                      2
Denmark                     2
Croatia                     2
Colombia                    2
Nigeria                     1
China                       1
Czech Republic              1
Philippines                 1
Japan     

In [45]:
df['state'].isna().value_counts()

False    742
True     515
Name: state, dtype: int64

In [46]:
#filling feature states by shifting NaN by "no state referred"
df['state'] = df['state'].fillna("no state referred")

In [47]:
df['state'].head()

0                   IL
1                   IN
2    no state referred
3    no state referred
4                   TX
Name: state, dtype: object

In [59]:
#Updating the cleaned data into the data set
Cleaned_data=df.to_csv('C:\\Users\\ARUN PRAVEEN\\Desktop\\mental health.csv')

In [60]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1257 entries, 0 to 1258
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   timestamp                  1257 non-null   datetime64[ns]
 1   country                    1257 non-null   object        
 2   state                      1257 non-null   object        
 3   self_employed              1257 non-null   object        
 4   family_history             1257 non-null   object        
 5   treatment                  1257 non-null   object        
 6   work_interfere             1257 non-null   object        
 7   no_employees               1257 non-null   object        
 8   remote_work                1257 non-null   object        
 9   tech_company               1257 non-null   object        
 10  benefits                   1257 non-null   object        
 11  care_options               1257 non-null   object        
 12  wellne

In [68]:
col=df.columns.value_counts()
print(col)
print("The total count of the columns is:",col.sum())

treatment                    1
care_options                 1
country                      1
mental_vs_physical           1
tech_company                 1
obs_consequence              1
supervisor                   1
timestamp                    1
coworkers                    1
family_history               1
phys_health_interview        1
work_interfere               1
leave                        1
benefits                     1
gender_new                   1
mental_health_consequence    1
self_employed                1
state                        1
no_employees                 1
phys_health_consequence      1
seek_help                    1
remote_work                  1
anonymity                    1
updated_age                  1
mental_health_interview      1
wellness_program             1
dtype: int64
The total count of the columns is: 26


As a conclusion , 
1)  I have done data cleaning such as imputing the NA values and filled null categorical variables,conversion to appropriate Data type.

2)  I have displayed the count of categorical variables in the dataset and percentage of null values (in the beginning) 